tests/parser/syntax/test_msg_data.py

```diff
@@ -0,0 +1,118 @@
+import pytest
+from eth_tester.exceptions import TransactionFailed
+
+from vyper import compiler
+from vyper.exceptions import StateAccessViolation, SyntaxException
+
+
+def test_variable_assignment(get_contract, keccak):
+    code = """
+@external
+def foo() -> Bytes[4]:
+    bar: Bytes[4] = slice(msg.data, 0, 4)
+    return bar
+"""
+
+    contract = get_contract(code)
+
+    assert contract.foo() == bytes(keccak(text="foo()")[:4])
+
+
+def test_slicing_start_index_other_than_zero(get_contract):
+    code = """
+@external
+def foo(_value: uint256) -> uint256:
+    bar: Bytes[32] = slice(msg.data, 4, 32)
+    return convert(bar, uint256)
+"""
+
+    contract = get_contract(code)
+
+    assert contract.foo(42) == 42
+
+
+def test_get_full_calldata(get_contract, keccak, w3):
```

Comment:
Would be good to include a test that has another statement following the assignment from `msg.data`, to ensure the free memory pointer is properly advanced.